vowpalwabbit/plt.cc

```diff
@@ -0,0 +1,382 @@
+        if (preds.label_v.size() >= p.top_k)
+          break;
+      }
+    }
+
+    // calculate p@
+    if (true_labels.size() > 0)
+    {
+      for (size_t i = 0; i < p.top_k; ++i)
+      {
+        if (true_labels.count(preds.label_v[i]))
+          ++p.tp_at[i];
+      }
+      ++p.ec_count;
+      p.true_count += true_labels.size();
+    }
+  }
+
+  ec.pred.multilabels = std::move(preds);
+  ec.l.multilabels = std::move(multilabels);
+}
+
+void finish_example(vw& all, plt& p, example& ec)
+{
+  MULTILABEL::output_example(all, ec);
+  VW::finish_example(all, ec);
+}
+
+void finish(plt& p)
+{
+  // print results in test mode
+  if (!p.all->training && p.ec_count > 0)
+  {
+    // top-k predictions
+    if (p.top_k > 0)
+    {
+      double correct = 0;
+      for (size_t i = 0; i < p.top_k; ++i)
+      {
+        correct += p.tp_at[i];
+        std::cerr << "p@" << i + 1 << " = " << correct / (p.ec_count * (i + 1)) << std::endl;
+        std::cerr << "r@" << i + 1 << " = " << correct / p.true_count << std::endl;
+      }
+    }
+
+    else if (p.threshold > 0)
+    {
+      std::cerr << "hamming loss = " << static_cast<double>(p.fp + p.fn) / p.ec_count << std::endl;
+      std::cerr << "precision = " << static_cast<double>(p.tp) / (p.tp + p.fp) << std::endl;
+      std::cerr << "recall = " << static_cast<double>(p.tp) / (p.tp + p.fn) << std::endl;
+    }
+  }
+}
+
+void save_load_tree(plt& p, io_buf& model_file, bool read, bool text)
+{
+  if (model_file.num_files() > 0)
+  {
+    bool resume = p.all->save_resume;
+    std::stringstream msg;
+    msg << ":" << resume << "\n";
+    bin_text_read_write_fixed(model_file, (char*)&resume, sizeof(resume), "", read, msg, text);
+
+    if (resume && !p.all->weights.adaptive)
+    {
+      for (size_t i = 0; i < p.t; ++i)
+        bin_text_read_write_fixed(model_file, (char*)&p.nodes_time[i], sizeof(p.nodes_time[0]), "", read, msg, text);
+    }
+  }
+}
+}  // namespace plt_ns
+
+using namespace plt_ns;
+
+base_learner* plt_setup(options_i& options, vw& all)
+{
+  auto tree = scoped_calloc_or_throw<plt>();
+  option_group_definition new_options("Probabilistic Label Tree ");
+  new_options.add(make_option("plt", tree->k).keep().help("Probabilistic Label Tree with <k> labels"))
+      .add(make_option("kary_tree", tree->kary).keep().default_value(2).help("use <k>-ary tree"))
+      .add(make_option("threshold", tree->threshold)
+               .default_value(0.5)
+               .help("predict labels with conditional marginal probability greater than <thr> threshold"))
+      .add(make_option("top_k", tree->top_k)
+               .default_value(0)
+               .help("predict top-<k> labels instead of labels above threshold"));
+  options.add_and_parse(new_options);
+
+  if (!options.was_supplied("plt"))
+    return nullptr;
+
+  tree->all = &all;
+
+  // calculate number of tree nodes
+  double a = std::pow(tree->kary, std::floor(std::log(tree->k) / std::log(tree->kary)));
```

Comment:
Brief comments describing what the intermediate (a,b,...e) values represent would be helpful

Reply:
They are used to divide the longer equation, it's hard to comment on them

Reply:
is this the equation to calculate number of total nodes (or just leaves) in a complete k-ary tree? or how would you describe it exactly, mostly curious rather than related to the PR

Reply:
Yes, the number of leaves (each leaf corresponds to one label) (`plt.k`) and tree arity (`plt.kary`) are arguments. Then this part calculates the total number of nodes and number of inner nodes (nodes without labels) in a complete k-ary tree for given arity and number of leaves. The tree is complete in the sense that each level of the tree is filled with nodes from left to right. As the results, it is possible to address parent and child nodes of a given node easily:
parent = floor(node_id - 1 / tree_arity)
first_child = node_id * tree_arity + 1
second_child = node_id * tree_arity + 2 ...